### PR TITLE
Support for functools.partial functions as callbacks

### PR DESCRIPTION
**Pull request by brendancol**
#### Description
This PR makes changes to callback validation found in `bokeh/util/callback_manager.py` and adds associated tests to `bokeh/util/tests/test_callback_manager.py`.

New additions which callback manager now accepts include:
 - partial functions: `partial(lambda w,x,y,z: x , 'partial value')` 
 - partial methods
 - functions with extra args with default values: `def f(attr, old, new, extra='default'):pass`
 - methods with extra args with default values

To run associated tests:

    py.test -q -s bokeh/util/tests/test_callback_manager.py

Nested partial objects are NOT currently valid.

### CODE REVIEW

**Comment by bryevdv** [comment]
I think we are stuck with `getargspec` and not `getfullargspec`

Also this might be useful:
```
In [12]: def foo(a, b, c=20): pass

In [13]: bar = partial(foo, 10)

In [14]: getargspec(bar.func)
Out[14]: ArgSpec(args=['a', 'b', 'c'], varargs=None, keywords=None, defaults=(20,))

In [15]: bar.args
Out[15]: (10,)
```
So `len(spec.args) - len(bar.args)` would yield `2` which is the correct number of free parameters.

File: bokeh/util/callback_manager.py [hunk 1]
```diff
@@ -17,6 +17,8 @@ def _callback_argspec(callback):
     elif ismethod(callback):
         # in this case the argspec will have 'self' in the args
         return getargspec(callback)
+    elif isinstance(callback, partial):
+        return getargspec(callback.func)
```
Comment (havocp) [inline bokeh/util/callback_manager.py]: the whole point of partial is that it changes the signature of `func` right?

It looks like `partial.__call__` may be implemented in native code? Then we hit the problem that native methods aren't introspectable: https://bugs.python.org/issue1748064

But maybe inspect.signature works ... python 3 only ? ... https://docs.python.org/3/library/inspect.html#inspect.signature  

I think what we might have to do is give up if the callback isn't a function or method, and `__call__` isn't a function or method either? would that screen out native calls?

File: bokeh/tests/test_callback_manager.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        k = Klass()
+        p = partial(k.callback2, check="test", check1="test1")
+        try:
+            _check_callback(p, self.expected_args)
+        except:
+            raise
+            self.fail("Callback check failed")
+
+    def test_raises_value_error_on_invalid_callback(self):
+        self.assertRaises(ValueError, _check_callback, lambda x: x, self.expected_args)
```
Comment (bryevdv) [inline bokeh/tests/test_callback_manager.py]: I'd encourage all new unit tests to be in `pytest` style.

Comment (bryevdv) [inline bokeh/tests/test_callback_manager.py]: Also shouldn't this be in `bokeh.util.tests` ?

Comment (brendancol) [inline bokeh/tests/test_callback_manager.py]: Yes, refactored and moved to `bokeh.util.tests`...

**Pull request merged**